### DISCUSSION TITLE
PostgREST-js optional filtering

### DISCUSSION THREAD

**Discussion by RickMeasham** [supabase/supabase#3312] [Questions] [answered]
How can I set up optional filtering on a query? Each of the filters is a distinct method call rather than an object I can build. The closest I can see is the `filter()` filter that would let me `.filter( columnName, operator, value )`. But to use that I need to know how many matches I need. I'm looking for something more like this, where I could have zero or more filters:

```
.filterList([
    { column: columnName, operator: operator, value: value },
    { column: columnName, operator: operator, value: value },
])
```

**Comment by GaryAustin1** [comment, marked as answer]
I just dealt with this by starting the query constructor with:
    let query = supabase.from('notes')
        .select("*")
        .eq('uid', uid)

then I loop thru a table like yours and do:
      query = query.op(column,value)   // op is eq, gt, textsearch....should be able to loop and add .filter(col,op,val) if they are simple filters

finally I do:
    let { data: notes, error } = await query
        .order('id', { ascending: false })

Seems to work, but I'm just beginning at SupaBase.  That was how I did it on Firebase.

**Reply by mstade** [reply]
Yup – we use this technique in various places, and I'd recommend it as a good way to set filters dynamically. The key is that the query isn't actually executed until the promise is created, so something like this works just fine:

```javascript
const query = supabase.from('wherever').select()

for (const filterParams of filterList) {
  query.filter(filterParams)
}

const { data, error } = await query // This executes the query
```

**Reply by RickMeasham** [reply]
This is a great solution thanks @GaryAustin1 and @mstade !

**Reply by devhandler** [reply]
supabase.from('wherever').select()
doesn't create a promise? 

i thought it created a promise just not awaited, executing in parallel and not blocking the following statement until it hits the await at the bottom.

**Reply by GaryAustin1** [reply]
You would need to add a .then() to the select to get it to execute and move on.

**Comment by soedirgo** [comment]
I'm not sure what you meant by "optional" filtering here, but in case you mean dynamically adding filters, you can try this: https://github.com/supabase/supabase/discussions/3080#discussioncomment-1282318
